libhttpserver/server.go

```diff
@@ -0,0 +1,170 @@
+type Server struct {
+	config libkbfs.Config
+	server *libkb.HTTPSrv
+
+	tokens *lru.Cache
+	fs     *lru.Cache
+}
+
+var _ libkbfs.LocalHTTPServer = (*Server)(nil)
+
+const tokenByteSize = 16
+
+// NewToken implements the libkbfs.LocalHTTPServer interface..
+func (s *Server) NewToken() (token string, err error) {
+	buf := make([]byte, tokenByteSize)
+	if _, err = rand.Read(buf); err != nil {
+		return "", err
+	}
+	token = hex.EncodeToString(buf)
+	s.tokens.Add(token, nil)
+	return token, nil
+}
+
+func (s *Server) handleInvalidToken(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusForbidden)
+	io.WriteString(w, `
+    <html>
+        <head>
+            <title>KBFS HTTP Token Invalid</title>
+        </head>
+        <body>
+            token invalid
+        </body>
+    </html>
+    `)
+}
+
+func (s *Server) handleBadRequest(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (s *Server) getHTTPFileSystem(ctx context.Context, requestPath string) (
+	toStrip string, fs http.FileSystem, err error) {
+	fields := strings.Split(requestPath, "/")
+	if len(fields) < 3 {
+		return "", nil, errors.New("bad path")
+	}
+
+	tlfType, err := tlf.ParseTlfType(fields[0])
+	if err != nil {
+		return "", nil, err
+	}
+
+	tlfHandle, err := libkbfs.GetHandleFromFolderNameAndType(ctx,
+		s.config.KBPKI(), s.config.MDOps(), fields[1], tlfType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	toStrip = path.Join(fields[0], fields[1])
+
+	fav := tlfHandle.ToFavorite()
+	if fsCached, ok := s.fs.Get(fav); ok {
+		if tlfFS, ok := fsCached.(*libfs.FS); ok {
+			return toStrip, tlfFS.ToHTTPFileSystem(ctx), nil
+		}
+	}
+
+	tlfFS, err := libfs.NewFS(ctx,
+		s.config, tlfHandle, "", "", keybase1.MDPriorityNormal)
+	if err != nil {
+		return "", nil, err
+	}
+	s.fs.Add(fav, tlfFS)
+
+	return toStrip, tlfFS.ToHTTPFileSystem(ctx), nil
+}
+
+// serve accetps "/<fs path>?token=<token>"
+// For example:
+//     /team/keybase/file.txt?token=1234567890abcdef1234567890abcdef
+func (s *Server) serve(w http.ResponseWriter, req *http.Request) {
+	token := req.URL.Query().Get("token")
+	if len(token) == 0 || !s.tokens.Contains(token) {
+		s.handleInvalidToken(w)
+		return
+	}
+	toStrip, fs, err := s.getHTTPFileSystem(req.Context(), req.URL.Path)
+	if err != nil {
+		s.handleBadRequest(w)
+		return
+	}
+	http.StripPrefix(toStrip, http.FileServer(fs)).ServeHTTP(w, req)
+}
+
+func overrideMimeType(ext, mimeType string) (newExt, newMimeType string) {
+	// Send text/plain for all HTML and JS files to avoid them being executed
+	// by the frontend WebView.
+	lower := strings.ToLower(mimeType)
+	if strings.Contains(lower, "javascript") ||
+		strings.Contains(lower, "html") {
+		return ext, "text/plain"
+	}
+	return ext, mimeType
+}
+
+const portStart = 7000
+const portEnd = 8000
+const requestPathRoot = "/files/"
+
+// Init implements the libkbfs.LocalHTTPServer interface.
+func (s *Server) Init(
```

Comment:
Maybe we should wrap the init code in a `sync.Once`, or somehow return an error if it's already initialized?